### PR TITLE
docs(js-toolkit): add documentation on how to test JS Toolkit locally in liferay-portal

### PR DESCRIPTION
**Pull request by izaera**
There are more ways to test this, but this one is the safest I've found and especially it works with
`ant all` too, which is not an easy thing because it tends to destroy any links or manual copies
you may have in your `node_modules` folder.

### CODE REVIEW

**Comment by izaera** [comment]
Just wanted to add that I have managed to make the bundler receive a different configuration than the one provided from npm-scripts even when it really knew that it should read it from the file provided by npm-scripts.

This has been after yarnlinking the css-loader alone and leaving the rest as usual. It seems that css-loader was getting some files from my js-toolkit development project and others from the portal which caused its internal modules to be in some kind of Schrödinger state.

I'm just saying it here to convince the world that the only repeatable, safe way to check local changes is by using a local npm registry (sadly). Otherwise it's impossible to know exactly what you are doing given the way node resolution works.

File: maintenance/projects/js-toolkit/CONTRIBUTING.md [hunk 1]
```diff
@@ -141,3 +141,44 @@ If you need to test local versions of the packages, you can install [Verdaccio](
 4. To stop using the local repository edit your `~/.npmrc` and `~/.yarnrc` files and remove the local repo.
 
 Publishing to the local Verdaccio repository won't update [https://npmjs.com](https://npmjs.com) so you can publish as many local versions as you want without worrying about polluting the public npm repository. Then, when you are finished testing, just remove the local versions from your local Verdaccio, point `npm` and `yarn` to the public npm repo, and publish the ultimate valid release.
+
+## Testing modifications in liferay-portal
+
+Testing in `liferay-portal` is a bit more difficult, because it doesn't use the JS Toolkit directly, but through `npm-scripts`. Because of that, if you want to test any modified piece of code in this repo within `liferay-portal`'s build you would need to first release a version of the JS Toolkit (which implies delivering new artifacts for every package), then update `npm-scripts`, the releasing a new version of them, then update the portal and test.
```
Comment (wincent) [inline maintenance/projects/js-toolkit/CONTRIBUTING.md:147]: "the releasing a new version" → "then release a new version"

File: maintenance/projects/js-toolkit/CONTRIBUTING.md [hunk 2]
```diff
@@ -141,3 +141,44 @@ If you need to test local versions of the packages, you can install [Verdaccio](
...
+
+Because this is quite unefficient, a more direct way is recommended using `yarn` resolutions and a local npm server (like [Verdaccio](https://verdaccio.org/)).
```
Comment (wincent) [inline maintenance/projects/js-toolkit/CONTRIBUTING.md:149]: "unefficient" → "inefficient"

File: maintenance/projects/js-toolkit/CONTRIBUTING.md [hunk 3]
```diff
@@ -141,3 +141,44 @@ If you need to test local versions of the packages, you can install [Verdaccio](
...
+	"resolutions": {
+		"liferay-npm-bundler-loader-css-loader": "2.24.2-alpha.0"
+	},
+
+	...
+
+}
+```
+
+Then make sure to be running `yarn` version 10 and execute `yarn` in `liferay-portal/modules` to update the `node_modules` directory. If you want to make sure you only have one copy of `liferay-npm-bundler-loader-css-loader` you can run `yarn why liferay-npm-bundler-loader-css-loader`.
```
Comment (wincent) [inline maintenance/projects/js-toolkit/CONTRIBUTING.md:171]: > `yarn` version 10

wat?

Did you mean Node?

File: maintenance/projects/js-toolkit/CONTRIBUTING.md [hunk 4]
```diff
@@ -141,3 +141,44 @@ If you need to test local versions of the packages, you can install [Verdaccio](
...
+After these steps you have a `liferay-portal` that will use your new version of css-loader even if doing `ant all`. Go ahead, test your changes and, when you are done simply set the npm registry back to the normal, remove the `resolutions` field, and proceed as usual.
+
+### What to do if you need to make more changes
+
+If you make more changes after testing, you have several options to update your `liferay-portal` project, but probably the easiest one is to release a new local version (say `2.24.2-alpha.1`, `...-alpha.2`, and so on) update the portal, and run `yarn` again. This is because:
+
+1. Verdaccio doesn't allow to republish an existing version unless you unpublish it first
+2. The portal has a local npm cache at `liferay-portal/modules/node_modules_cache` and it will complain if you try to use a modified package with the same version number
+3. Yarn has its own cache than needs to be cleaned too if you want to reuse version numbers
+4. The `yarn.lock` file saves package checksums and the portal's build complains if you try to reuse a version number
```
Comment (wincent) [inline maintenance/projects/js-toolkit/CONTRIBUTING.md:182]: So many helpful measures working together! ❤️

File: maintenance/projects/js-toolkit/CONTRIBUTING.md [hunk 5]
```diff
@@ -141,3 +141,44 @@ If you need to test local versions of the packages, you can install [Verdaccio](
...
+
+All these things may be circumvented (in fact I have scripts to do so) but I don't think it makes any sense documenting such a hacky and error prone process here, since it is much simpler to change the version number and let things flow normally.
```
Comment (wincent) [inline maintenance/projects/js-toolkit/CONTRIBUTING.md:184]: "error prone" → "error-prone"

**Review by wincent — APPROVED** [review]
lg2m

> Otherwise it's impossible to know exactly what you are doing given the way node resolution works.

Not sure "impossible" is the right word there. Think you meant "hard". 😉

File: maintenance/projects/js-toolkit/CONTRIBUTING.md (hunk 3 above)
Comment (izaera) [reply maintenance/projects/js-toolkit/CONTRIBUTING.md:171]: Ah yes. Node :sweat_smile:

File: maintenance/projects/js-toolkit/CONTRIBUTING.md (hunk 4 above)
Comment (izaera) [reply maintenance/projects/js-toolkit/CONTRIBUTING.md:182]: Nevertheless, I have scripts to defeat them all. But don't tell anybody :no_mouth:

**Comment by izaera** [comment]
Fixed the errors and force pushed to rewrite commit title

**Comment by izaera** [comment]
I'm merging this since it is approved and green.

**Pull request merged**